src/Features/Core/Portable/Structure/Syntax/AbstractBlockStructureProvider.cs

```diff
@@ -67,9 +68,9 @@ public override async Task ProvideBlockStructureAsync(BlockStructureContext cont
         private void ProvideBlockStructureWorker(
             BlockStructureContext context, SyntaxNode syntaxRoot)
         {
-            using var _ = ArrayBuilder<BlockSpan>.GetInstance(out var spans);
+            using var spans = TemporaryArray<BlockSpan>.Empty;
             BlockSpanCollector.CollectBlockSpans(
-                context.Document, syntaxRoot, _nodeProviderMap, _triviaProviderMap, spans, context.CancellationToken);
+                context.Document, syntaxRoot, _nodeProviderMap, _triviaProviderMap, ref Unsafe.AsRef(in spans), context.CancellationToken);
```

Comment:
so i'm not opposed to this.  but this feels like it would not actually be an perf improvement.  the reason being that near all files will have +4 elements in them.  as such, we'll always spill to the builder, so we'll only be adding overhead here.

That said, i'm ok with this due to consistency if we just want to use this almost anywhere, accepting that in some cases we'll have a little extra perf hit.  One thing that i would think could be interesting (outside of the scope of this pr) is if we ended up having thread-local pools for our underlying builders. That way we would never have a single problem colliding when we returned the builders.

Reply:
> if we ended up having thread-local pools

Once we switch to .NET Core, I plan to rewrite the pool to have _processor core_-specific first elements hinted by [`Thread.GetCurrentProcessorId`](https://docs.microsoft.com/en-us/dotnet/api/system.threading.thread.getcurrentprocessorid?view=netcore-3.1). This will have performance similar to thread local storage but without the corresponding working set increase. `TemporaryArray<T>` would still be better for small storage since the pool efficiency is reduced as the number of active leases increases.